File: updraft_sdk/src/main/java/com/apsswithlove/updraft_sdk/manager/AppUpdateManager.java
package com.apsswithlove.updraft_sdk.manager;

import android.app.Application;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;
import android.arch.lifecycle.ProcessLifecycleOwner;
import com.apsswithlove.updraft_sdk.interactor.CheckUpdateInteractor;
import com.apsswithlove.updraft_sdk.presentation.UpdraftSdkUi;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by satori on 3/27/18.
 */

public class AppUpdateManager implements LifecycleObserver, UpdraftSdkUi.Listener {

    private CheckUpdateInteractor mCheckUpdateInteractor;
    private UpdraftSdkUi mUpdraftSdkUi;
    private Disposable mCheckUpdateDisposable;

    public AppUpdateManager(Application application,
                            CheckUpdateInteractor checkUpdateInteractor,
                            UpdraftSdkUi updraftSdkUi) {
        mCheckUpdateInteractor = checkUpdateInteractor;
        mUpdraftSdkUi = updraftSdkUi;
    }

    public void start() {
        ProcessLifecycleOwner.get().getLifecycle().addObserver(this);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void onAppForegrounded() {
        mUpdraftSdkUi.showStartAlert();
        mUpdraftSdkUi.setListener(this);
        if (mCheckUpdateDisposable != null) {
            mCheckUpdateDisposable.dispose();
        }
        mCheckUpdateDisposable = mCheckUpdateInteractor
                .checkUpdate()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        checkUpdateResultModel -> {
                            if (checkUpdateResultModel.isShowAlert()) {
                                mUpdraftSdkUi.showAlert(checkUpdateResultModel.getUrl());
                            }
                        },
                        Throwable::printStackTrace
                );
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    public void onAppBackgrounded() {
        mUpdraftSdkUi.setListener(null);
        if (mCheckUpdateDisposable != null) {
            mCheckUpdateDisposable.dispose();
        }
    }

    @Override
    public void onOkClicked(String url) {
        mUpdraftSdkUi.openUrl(url);
    }
}

File: updraft-sdk/src/main/java/com/appswithlove/updraft/manager/AppUpdateManager.kt
package com.appswithlove.updraft.manager

import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.lifecycleScope
import com.appswithlove.updraft.interactor.CheckUpdateInteractor
import com.appswithlove.updraft.presentation.UpdraftSdkUi
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

/**
 * Created by satori on 3/27/18.
 */
class AppUpdateManager(
    private val checkUpdateInteractor: CheckUpdateInteractor,
    private val updraftSdkUi: UpdraftSdkUi
) : DefaultLifecycleObserver, UpdraftSdkUi.Listener {

    private var checkUpdateJob: Job? = null

    fun start() {
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
    }

    override fun onStart(owner: LifecycleOwner) {
        updraftSdkUi.showFeedbackAlert()
        updraftSdkUi.setListener(this)

        checkUpdateJob?.cancel()
        checkUpdateJob = owner.lifecycleScope.launch {
            try {
                val result = checkUpdateInteractor.checkUpdate()
                val url = result.url
                if (result.showAlert && url != null) {
                    updraftSdkUi.showUpdateAvailableAlert(url)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun onStop(owner: LifecycleOwner) {
        updraftSdkUi.setListener(null)
        checkUpdateJob?.cancel()
    }

    override fun onOkClicked(url: String) {
        updraftSdkUi.openUrl(url)
    }
}
